### PR TITLE
Partition the Pauli group into qubit-wise commuting terms

### PR DESCRIPTION
**Pull request by trbromley**
**Context:**

As part of the QCut module, we need to efficiently measure over the full N-qubit Pauli group. Although there are `4^N` terms, we can measure only `3^N` groups.

**Description of the Change:**

Added the `partition_pauli_group` function to the `grouping` module.

**Possible Drawbacks:**

The function returns Pauli words expressed as simple strings. This is sufficient for the QCut module's needs but some users may prefer the Pauli words as PL observables/tensors. Since the current implementation is sufficient for now, I propose to leave returning PL observables/tensors as a follow up addition.

Example:
```pycon
>>> qml.grouping.partition_pauli_group(3)
[['III', 'IIZ', 'IZI', 'IZZ', 'ZII', 'ZIZ', 'ZZI', 'ZZZ'],
 ['IIX', 'IZX', 'ZIX', 'ZZX'],
 ['IIY', 'IZY', 'ZIY', 'ZZY'],
 ['IXI', 'IXZ', 'ZXI', 'ZXZ'],
 ['IXX', 'ZXX'],
 ['IXY', 'ZXY'],
 ['IYI', 'IYZ', 'ZYI', 'ZYZ'],
 ['IYX', 'ZYX'],
 ['IYY', 'ZYY'],
 ['XII', 'XIZ', 'XZI', 'XZZ'],
 ['XIX', 'XZX'],
 ['XIY', 'XZY'],
 ['XXI', 'XXZ'],
 ['XXX'],
 ['XXY'],
 ['XYI', 'XYZ'],
 ['XYX'],
 ['XYY'],
 ['YII', 'YIZ', 'YZI', 'YZZ'],
 ['YIX', 'YZX'],
 ['YIY', 'YZY'],
 ['YXI', 'YXZ'],
 ['YXX'],
 ['YXY'],
 ['YYI', 'YYZ'],
 ['YYX'],
 ['YYY']]
```

### CODE REVIEW

**Comment by trbromley** [comment]
[sc-14640]

File: pennylane/grouping/pauli.py [hunk 1]
```diff
@@ -250,3 +253,80 @@ def pauli_mult_with_phase(pauli_1, pauli_2, wire_map=None):
             phase *= -1j
 
     return pauli_product, phase
+
+
+@lru_cache()
```
Comment (anthayes92) [inline pennylane/grouping/pauli.py:258]: how did you know to use this decorator here? I don't really understand how caching works and when to take advantage of it!

File: tests/grouping/test_pauli_group.py [hunk 2]
```diff
@@ -196,3 +202,58 @@ def test_pauli_mult_with_phase(self, pauli_word_1, pauli_word_2, wire_map, expec
...
+            ["YXI", "YXZ"],
+            ["YXX"],
+            ["YXY"],
+            ["YYI", "YYZ"],
+            ["YYX"],
+            ["YYY"],
+        ]
+        assert expected == partition_pauli_group(3)
+
+    @pytest.mark.parametrize("n", range(1, 9))
```
Comment (anthayes92) [inline tests/grouping/test_pauli_group.py:243]: for the case of `partition_pauli_group(0)` I get `[['']]` but maybe an exception would be appropriate?

**Review by anthayes92 — APPROVED** [review]
Looks awesome, thanks @trbromley ! 

I've left a minor Q and suggestion. Approved :heavy_check_mark:

File: pennylane/grouping/pauli.py (hunk 1 above)
Comment (trbromley) [reply pennylane/grouping/pauli.py:258]: There's probably better explanations out there, but from my perspective `partition_pauli_group` fits the criteria nicely of:
- simple inputs that can be hashed
- function takes a long time to compute
- restricted set of inputs that may be called multiple times (e.g., in circuit cutting we might have a few fragments with 3-wire measure nodes)

File: tests/grouping/test_pauli_group.py (hunk 2 above)
Comment (trbromley) [reply tests/grouping/test_pauli_group.py:243]: Added some validation here: 7b804bd52818c73bcca60009ec0eace02066f277

**Review by trbromley — COMMENTED** [review]
Thanks @anthayes92!

File: tests/grouping/test_pauli_group.py [hunk 3]
```diff
@@ -196,3 +202,66 @@ def test_pauli_mult_with_phase(self, pauli_word_1, pauli_word_2, wire_map, expec
...
+            for i in range(size):
+                for j in range(i, size):
+                    s1 = group[i]
+                    s2 = group[j]
+                    w1 = string_to_pauli_word(s1)
+                    w2 = string_to_pauli_word(s2)
+                    assert is_commuting(w1, w2)
+
+    def test_pauli_group_invalid_input(self):
+        """Test that invalid inputs to the Pauli group are handled correctly."""
```
Comment (trbromley) [inline tests/grouping/test_pauli_group.py:262]:
```suggestion
    def test_invalid_input(self):
        """Test that invalid inputs are handled correctly."""
```

**Pull request merged**